src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs

```diff
@@ -895,33 +898,66 @@ private static IEnumerable<PSTypeName> InferTypesFromCimCommand(PseudoBindingInf
             yield return new PSTypeName(typeof(CimInstance));
         }
 
-        private IEnumerable<PSTypeName> InferTypesFromForeachCommand(PseudoBindingInfo pseudoBinding)
+        private IEnumerable<PSTypeName> InferTypesFromForeachCommand(PseudoBindingInfo pseudoBinding, CommandAst commandAst)
         {
-            AstParameterArgumentPair argument;
-            if (pseudoBinding.BoundArguments.TryGetValue("Begin", out argument))
+            var previousPipelineElement = GetPreviousPipelineCommand(commandAst);
+            if (previousPipelineElement == null)
             {
-                foreach (var type in GetInferredTypeFromScriptBlockParameter(argument))
-                {
-                    yield return type;
-                }
+                yield break;
             }
-
-            if (pseudoBinding.BoundArguments.TryGetValue("Process", out argument))
+            foreach (var t in InferTypes(previousPipelineElement))
             {
-                foreach (var type in GetInferredTypeFromScriptBlockParameter(argument))
+                var thisToRestore = _context.CurrentThisType;
+                try
                 {
-                    yield return type;
-                }
-            }
+                    _context.CurrentThisType = t;
```

Comment:
Agree about the test, but it will require some refactoring to pass the type down to the scriptblock handling. Worth it?